File: backend/app/schemas/ai.py
from __future__ import annotations

from typing import Any

from pydantic import BaseModel, ConfigDict, Field


class RunSummary(BaseModel):
    run_id: str
    status: str | None = None
    model_name: str | None = None
    timeframe: str | None = None
    horizon: int | None = None
    created_at: str | None = None
    model_ver: str | None = None
    feature_version: str | None = None
    band_mode: str | None = None
    checkpoint_path: str | None = None
    best_epoch: int | None = None
    best_val_total: float | None = None
    line_target_type: str | None = None
    band_target_type: str | None = None


class RunDetail(RunSummary):
    val_metrics: dict[str, Any] = Field(default_factory=dict)
    test_metrics: dict[str, Any] = Field(default_factory=dict)
    config_summary: dict[str, Any] = Field(default_factory=dict)
    wandb_run_id: str | None = None
    config: dict[str, Any] | None = None

    model_config = ConfigDict(extra="allow")


class EvaluationSummary(BaseModel):
    run_id: str
    ticker: str | None = None
    timeframe: str | None = None
    asof_date: str | None = None
    coverage: float | None = None
    lower_breach_rate: float | None = None
    upper_breach_rate: float | None = None
    avg_band_width: float | None = None
    direction_accuracy: float | None = None
    mae: float | None = None
    smape: float | None = None
    spearman_ic: float | None = None
    top_k_long_spread: float | None = None
    top_k_short_spread: float | None = None
    long_short_spread: float | None = None
    fee_adjusted_return: float | None = None
    fee_adjusted_sharpe: float | None = None
    fee_adjusted_turnover: float | None = None


class BacktestSummary(BaseModel):
    run_id: str
    strategy_name: str | None = None
    timeframe: str | None = None
    return_pct: float | None = None
    sharpe: float | None = None
    mdd: float | None = None
    win_rate: float | None = None
    profit_factor: float | None = None
    num_trades: int | None = None
    fee_adjusted_return_pct: float | None = None
    fee_adjusted_sharpe: float | None = None
    avg_turnover: float | None = None
    meta: dict[str, Any] = Field(default_factory=dict)
    created_at: str | None = None

File: frontend/src/api/types/ai.ts
import type { DisplayTimeframe } from "./common";

export type AiRunStatus = "completed" | "failed_nan" | "failed_quality_gate";

export interface AiRunSummary {
  run_id: string;
  status: AiRunStatus | string | null;
  model_name: string | null;
  timeframe: DisplayTimeframe | string | null;
  horizon: number | null;
  created_at: string | null;
  model_ver: string | null;
  feature_version: string | null;
  band_mode: string | null;
  checkpoint_path: string | null;
  best_epoch: number | null;
  best_val_total: number | null;
  line_target_type: string | null;
  band_target_type: string | null;
  role: string | null;
  feature_set: string | null;
  checkpoint_selection: string | null;
  wandb_status: string | null;
  deprecated_for_phase1_product_contract: boolean | string | null;
  indicator_layer_replacement: string | null;
  is_legacy: boolean;
}

export interface AiRunDetail extends AiRunSummary {
  val_metrics: Record<string, unknown>;
  test_metrics: Record<string, unknown>;
  config_summary: Record<string, unknown>;
  wandb_run_id: string | null;
  config?: Record<string, unknown> | null;
}

export interface EvaluationSummary {
  run_id: string;
  ticker: string | null;
  timeframe: string | null;
  asof_date: string | null;
  coverage: number | null;
  avg_band_width: number | null;
  direction_accuracy: number | null;
  mae: number | null;
  smape: number | null;
  spearman_ic: number | null;
  top_k_long_spread: number | null;
  top_k_short_spread: number | null;
  long_short_spread: number | null;
  fee_adjusted_return: number | null;
  fee_adjusted_sharpe: number | null;
  fee_adjusted_turnover: number | null;
}

export interface BacktestSummary {
  run_id: string;
  strategy_name: string | null;
  timeframe: string | null;
  return_pct: number | null;
  sharpe: number | null;
  mdd: number | null;
  win_rate: number | null;
  profit_factor: number | null;
  num_trades: number | null;
  fee_adjusted_return_pct: number | null;
  fee_adjusted_sharpe: number | null;
  avg_turnover: number | null;
  meta: Record<string, unknown>;
  created_at: string | null;
}
